host_core/lib/host_core/web_assembly/imports.ex

```diff
@@ -121,132 +121,148 @@ defmodule HostCore.WebAssembly.Imports do
-          actor,
-          binding,
-          namespace,
-          operation,
-          payload,
-          claims,
-          seed,
-          prefix,
-          provider_key,
-          state
-        )
-        |> finish_host_call(agent)
-
-      # Link definition not found for actor, could be attempting to call a call alias
-      :error ->
-        case lookup_call_alias(namespace) do
-          {:ok, actor_key} ->
-            finish_host_call(
-              {:ok, actor_key, binding, namespace, operation, payload, claims, seed, prefix,
-               provider_key, state},
-              agent
-            )
-
-          :error ->
-            Logger.error("Failed to find link definition or call alias for invocation")
-            0
-        end
-    end
+    %{token | target: target}
+  end
+
+  defp verify_link(token = %{target: :unknown}) do
+    %{token | verified: false}
   end
 
-  defp authorize(
-         actor,
-         binding,
-         namespace,
-         operation,
-         payload,
-         claims,
-         seed,
-         prefix,
-         provider_key,
-         state
+  defp verify_link(token = %{target: {:actor, _, _}}) do
+    %{token | verified: true}
+  end
+
+  defp verify_link(
+         token = %{
+           target: {:provider, _pk, _topic},
+           source_actor: actor,
+           namespace: namespace,
+           binding: binding
+         }
        ) do
-    if Enum.member?(claims.caps, namespace) do
-      {:ok, actor, binding, namespace, operation, payload, claims, seed, prefix, provider_key,
-       state}
-    else
-      {:error, "Actor unauthorized"}
-    end
+    verified =
+      case HostCore.Linkdefs.Manager.lookup_link_definition(actor, namespace, binding) do
+        {:ok, _ld} -> true
+        _ -> false
+      end
+
+    %{token | verified: verified}
   end
 
-  defp lookup_call_alias(call_alias) do
-    case :ets.lookup(:callalias_table, call_alias) do
-      [{_call_alias, pkey}] ->
-        {:ok, pkey}
+  defp authorize_call(token = %{verified: false}) do
+    %{token | authorized: false}
+  end
 
-      [] ->
-        :error
-    end
+  defp authorize_call(token = %{target: {:actor, _pk, _topic}}) do
+    %{token | authorized: true}
   end
 
-  defp finish_host_call({:error, reason}, agent) do
-    Agent.update(agent, fn state -> %State{state | host_error: reason} end)
+  defp authorize_call(
+         token = %{target: {:provider, _pk, _topic}, claims: claims, namespace: namespace}
+       ) do
+    %{token | authorized: Enum.member?(claims.caps, namespace)}
+  end
+
+  defp invoke(_token = %{authorized: false, agent: agent}) do
+    Agent.update(agent, fn state -> %State{state | host_error: "Invocation not authorized"} end)
     0
   end
 
-  defp finish_host_call(
-         {:ok, actor, binding, namespace, operation, payload, _claims, seed, prefix, provider_key,
-          state},
-         agent
+  defp invoke(
+         _token = %{
+           authorized: true,
```

Comment:
I struggled with this when refactoring, is there any variation of function names that makes `invoke` and `do_invoke` less confusing? I like this better than `finish_host_call` and `invoke`, 🤔 

Thoughts on the top-level pipeline to be:
```
token
    |> identify_target()
    |> verify_link()
    |> authorize_call()
    |> generate_invocation()
    |> invoke()
```

Just a brainstorm. If this feels worse to you then I'm happy with where it is

Reply:
refactored the names here